src/parser/utils.rs

```diff
@@ -558,53 +558,39 @@ pub fn mk_label(types: Types, src_id: FileId, l: usize, r: usize) -> Label {
     }
 }
 
-/// Generate a `Let` or a `LetPattern` (depending on `pat` being empty or not) from the parsing of
-/// a let definition. This function fails if the definition has both a non-empty pattern and is
-/// recursive (`pat != Destruct::Empty && rec`), because recursive let-patterns are currently not
-/// supported.
+/// Generate a `Let` or a `LetPattern` (depending on whether `assgn` has a record pattern) from
+/// the parsing of a let definition. This function fails if the definition has both a pattern
+/// and is recursive because recursive let-patterns are currently not supported.
 pub fn mk_let(
     rec: bool,
-    id: Option<Ident>,
-    pat: Option<RecordPattern>,
+    assgn: FieldPattern,
     t1: RichTerm,
     t2: RichTerm,
     span: RawSpan,
 ) -> Result<RichTerm, ParseError> {
-    let result = match pat {
-        Some(d) => {
-            if rec {
-                return Err(ParseError::RecursiveLetPattern(span));
-            }
-            mk_term::let_pat(id, d, t1, t2)
-        }
-        None => {
-            if let Some(id) = id {
-                if rec {
-                    mk_term::let_rec_in(id, t1, t2)
-                } else {
-                    mk_term::let_in(id, t1, t2)
-                }
-            } else {
-                panic!("unexpected let-binding without pattern or identifier")
-            }
+    match assgn {
+        FieldPattern::Ident(id) => Ok(if rec {
+            mk_term::let_rec_in(id, t1, t2)
+        } else {
+            mk_term::let_in(id, t1, t2)
+        }),
```

Comment:
Leave the decision to you, but IMHO, I find guarded patterns to be slightly more readable:

```rust
        FieldPattern::Ident(id) if rec => Ok(mk_term::let_rec_in(id, t1, t2)),
        FieldPattern::Ident(id) => Ok(mk_term::let_in(id, t1, t2)),
```